### PR TITLE
Provision new production Bertly environment.

### PR DESCRIPTION
**Pull request by DFurnes**
This pull request does some prep-work to create and then migrate to a new production Bertly stack (since we're currently running traffic off of the older `bertly-dev-app` stack in our prod organization).

### CODE REVIEW

**Comment by DFurnes** [comment]
Provisioned a new production stack under our [new production AWS organization](https://github.com/DoSomething/devops/issues/420). This environment uses the same API key, but has newly provisioned RDS & ElastiCache instances (and no custom domain):

```
https://mr0lbb7sak.execute-api.us-east-1.amazonaws.com/prod
```

**Comment by DFurnes** [comment]
So for syncing over data to the new instance, it looks like we've got two tasks:

1. Synchronize the PostgreSQL instances using [AWS Database Migration Service](https://console.aws.amazon.com/dms/home). We already hav the current production Bertly Compose.io instance set up (for Quasar), so we'll just need to add the new one (and to do so, allow our root account to access the new Bertly production VPC & create a replication subnet group).
2. Synchronize the ElastiCache instance, which we should be able to do by creating the new production instance with `SnapshotArns` set to a `*.rdb` dump in S3. From what I can tell, we can't use Redis replication w/ a Redis server outside of our AWS VPC, so this will have to involve a teeny bit of downtime.

**Comment by sheyd** [comment]
@DFurnes Not sure if a type or not, but we're using RDS for Bertly PostgreSQL instance. If we have a Compose.io one, it should be de-comm'd. I'm happy to sit with you if you'd like a 2nd set of eyes/hands on this. Also, for the continuing sync of Bertly data to Quasar, I need to run some RDS commands on the new Bertly instance.

**Comment by DFurnes** [comment]
Oh, yep got that wrong. Updated my comment! 👍

**Comment by mshmsh5000** [comment]
OK, so we'd like to run an experiment to discover CloudFormation behavior when importing an `rdb` file:

- Upload `dump.rdb` from the Compose instance to an S3 bucket in the new AWS production account
- Add the ARNS for this file to CloudFormation as the `SnapshotArns` param for the Elasticache instance
- Reprovision
- Verify that the Redis data got imported into the new instance
- Remove the `SnapshotArns` value
- Reprovision
- Observe what CloudFormation does: Does it leave alone the instance & data we provisioned, or does it delete that instance / wipe the data and send us back to square one?

**Comment by mshmsh5000** [comment]
DMS across accounts: https://aws.amazon.com/blogs/database/set-up-aws-dms-for-cross-account-migration/

**Comment by DFurnes** [comment]
So, progress update -

We added `SnapshotArns` to the CloudFormation template in e0ad7cb, and set the `/bertly/prod/redis-snapshot` variable to `arn:aws:s3:::bertly-prod-serverlessdeploymentbucket-pn2ydi42f3mw/serverless/bertly/prod/redis/dump.rdb` in SSM.

Deployed 16:48, started downloading S3 snapshot at 16:52, then failed with a relatively unhelpful "Failed to retrieve file from S3" at 16:53... meanwhile, our CloudFormation stack is still stuck at `UPDATE_IN_PROGRESS` and happily skipping along 30 minutes later. Rut roh.

**Comment by DFurnes** [comment]
Matt created a new `bertly-redis` S3 bucket yesterday and uploaded the same `dump.rdb` there (in case there was some weirdness loading the auto-provisioned bucket we'd used above). Unfortunately, we seem to have run into the same error (again, approx. ~1 minute after it started to fetch the S3 file):

```
Download of snapshot arn:aws:s3:::bertly-redis/dump.rdb failed for node 0001. Failed to retrieve file from S3
```

The new Redis instance is once again marked with a status of `restore-failed` and the CloudFormation seems to be stalled at `UPDATE_IN_PROGRESS` for that instance.

**Comment by DFurnes** [comment]
For the heck of it, I tried setting public read permissions on the `bertly-redis` bucket and that `dump.rdb` file (clearly not a long term solution) and it worked! So there's got to be some sort of permissioning issue between the `serverless-admin-prod` user and that bucket...

```
Wed Jun 27 11:11:10 GMT-400 2018 | This cache cluster does not support persistence (ex: 'appendonly'). Please use a different instance type to enable persistence.
Wed Jun 27 11:11:10 GMT-400 2018 | Cache cluster created
Wed Jun 27 11:11:08 GMT-400 2018 | Restore from arn:aws:s3:::bertly-redis/dump.rdb succeeded for node 0001
```

One other note from the logs - it looks like we may need to change an additional setting to make sure this instance persists data between restarts. Here's what I've found on [persistence in ElastiCache Redis](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/RedisAOF.html).

**Comment by DFurnes** [comment]
Hah, it looks like you need to [grant permissions to AWS itself to read that file](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/backups-seeding-redis.html#backups-seeding-redis-grant-access) (right right, of course 🙃). I've removed public read access to that bucket/file, granted the specific account permissions from those instructions, and trying a fresh re-provision.

**Comment by DFurnes** [comment]
Alrighty, that seems to have worked this time! Good stuff.

I've enabled daily backups for this instance (which doesn't require a restart or re-provision) & confirmed that the data on the Redis instance was not overwritten with the earlier S3 snapshot. 👌 Now to see how we deal with restarts...

**Comment by DFurnes** [comment]
Welp, I did a little digging into Redis persistence and it looks like `appendonly` (as suggested in that log message above & that documentation page I found) is [no longer modifiable](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/ParameterGroups.Redis.html#ParameterGroups.Redis.3-2-4.Changed) on AWS! According to [this FAQ](https://aws.amazon.com/elasticache/redis/faqs/), the suggested persistence method is now to just rely on daily snapshots:

> Yes, you can achieve persistence by snapshotting your Redis data using the Backup and Restore feature. Please see [here](http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/ManagingReplication.html) for details.

🤷‍♂️

**Comment by mshmsh5000** [comment]
Eff. That is less than optimal.

**Comment by DFurnes** [comment]
Yeah, not great at all. Just to double-check, I tried creating a link & restarting the ElastiCache instance - zap! All gone. And like, actually _all_ gone - including links made yesterday which should have been included in our nightly snapshots & things that were seeded in the initial S3 snapshot.

It sounds like the same thing would happen if we hit our memory limit for the instance (which to be fair, makes fine enough sense for an "elastic cache" but definitely falls into the "less than optimal" bucket for what we're trying to do).

**Comment by DFurnes** [comment]
I think we'll move forward assuming that we're connecting to an external [Compose.io](http://compose.io) Redis instance since ElastiCache doesn't handle persistence (and we need to add support for logging user-agents to production soon-ish).

Once we get things stabilized on production, we can evaluate other options to make this less manual.

**Comment by DFurnes** [comment]
A fun new issue popped up: it seems like once we've created a custom VPC for this stack (in order to hook up the Lambda function to the RDS instance), we no longer have internet access (for Compose). According to [Amazon's documentation](https://docs.aws.amazon.com/lambda/latest/dg/vpc.html#vpc-internet):

> Therefore, if your Lambda function requires Internet access (for example, to access AWS services that don't have VPC endpoints ), you can configure a NAT instance inside your VPC or you can use the Amazon VPC NAT gateway.

Great great, nice. I've pushed up a WIP attempt at that in 01f9c1e and will continue poking around with it in the morning since the rest of my day is booked.

File: bertly.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import redis
 import short_url
 import hashlib
+import httplib
```
[the remark this answers is not in the document]
Comment (DFurnes) [reply bertly.py]: Gah! Leftover from testing network requests. Removed! 🗑

**Comment by DFurnes** [comment]
Alright, this is ready for review I think. I've deployed this branch to the `dev` stage & tested that the shortlinks are correctly stashed in Compose & clicks are tracked in the CloudFormation-provisioned RDS instance.

**Comment by DFurnes** [comment]
I think these should be the remaining steps once this is merged:

1. Use Database Migration Service to get data from our current PostgreSQL instance into this new CloudFormation-provisioned one in the new production organization.
2. Hook up that new PostgreSQL instance to Quasar so our data pipeline stays functional.
3. Swap over our domain to the new prod organization & un-comment [this line](https://github.com/DoSomething/bertly/pull/34/files#diff-220045f50daee3e35e21b9d5fe19d67eR24).

**Comment by DFurnes** [comment]
Did a little more tidying up in b83d5e1, and deployed to dev & QA stages. It's worth noting that some of the changes, particularly to subnets, required deleting & re-creating the CloudFormation stack:

- **Development:** https://h4hpyr1pul.execute-api.us-east-1.amazonaws.com/dev ([bertly-dev-app](https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/bertly-dev-app))
- **QA:** https://a6xomad00m.execute-api.us-east-1.amazonaws.com/qa  ([bertly-qa-app](https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/bertly-qa-app))

As before, migrations must be [manually invoked](https://github.com/DoSomething/bertly/issues/21#issuecomment-396733920) when necessary.

**Comment by mshmsh5000** [comment]
What a long, strange trip it's been. This looks great.

We'll have to be careful coordinating step three (moving the domain to the new prod account), as it involves [interacting with AWS customer service](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/domain-transfer-between-aws-accounts.html). If the entire deployment were ready to go with updated data, then we wouldn't be concerned—the traffic would start routing to the new deployment, and it would produce the same responses.

My main concern here is that IIRC the Serverless domain setup procedure won't complete unless the domain it's attaching to is actually set up with a cert in Route 53. If that's the case, we can't run the [domain manager setup](https://www.npmjs.com/package/serverless-domain-manager#how-it-works) until customer service confirms that the transfer is complete.

Question for another issue: Is there a way we can do this without some unknown period of downtime?

File: serverless.yml [hunk 2]
```diff
@@ -67,30 +63,90 @@ resources:
...
+    BertlyPrivateSubnetB:
       DependsOn: BertlyVPC
       Type: AWS::EC2::Subnet
       Properties:
         VpcId:
           Ref: BertlyVPC
-        AvailabilityZone: ${self:provider.region}d
-        CidrBlock: '10.0.2.0/24'
+        AvailabilityZone: ${self:provider.region}b
+        CidrBlock: '10.0.3.0/24'
```
Comment (mshmsh5000) [inline serverless.yml]: This is a subtle change. Can you explain the change from `2` to `3` in that octet?

Comment (mshmsh5000) [reply serverless.yml]: Nevermind! Got it. Because you have three subnets now.

File: serverless.yml [hunk 3]
```diff
@@ -107,23 +163,18 @@ resources:
...
     BertlyRDSSubnetGroup:
       Type: AWS::RDS::DBSubnetGroup
       Properties:
         DBSubnetGroupDescription: 'RDS Subnet Group'
         SubnetIds:
-        - Ref: BertlySubnetA
-        - Ref: BertlySubnetB
-        - Ref: BertlySubnetC
+        - Ref: BertlyPrivateSubnetA
+        - Ref: BertlyPrivateSubnetB
```
Comment (mshmsh5000) [inline serverless.yml]: Ah, so here's my actual subnet question. Why are there two private subnets?

Comment (DFurnes) [reply serverless.yml]: Amazon requires that RDS instances [have a subnet group with at least two subnets in two different availability zones](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbsubnet-group.html). My guess is that's the minimum for them to ensure their availability guarantees?

**Comment by DFurnes** [comment]
> Question for another issue: Is there a way we can do this without some unknown period of downtime?

Hmm, yeah that seems tricky. 😖 We should definitely be able to get the data all synced up between the old and new production instances, but it does seem like we might have some downtime while the domain transfers orgs... I'll do some digging!

File: serverless.yml (hunk 3 above)
Comment (mshmsh5000) [reply serverless.yml]: Got it! 👍

**Pull request merged**